src/gui/editorwidgets/qgsdoublespinbox.cpp

```diff
@@ -140,13 +146,24 @@ void QgsDoubleSpinBox::setLineEditAlignment( Qt::Alignment alignment )
   mLineEdit->setAlignment( alignment );
 }
 
+void QgsDoubleSpinBox::setSpecialValueText( const QString &txt )
+{
+  if ( txt.isEmpty() )
+    QDoubleSpinBox::setSpecialValueText( SPECIAL_TEXT_WHEN_EMPTY );
+  else
+    QDoubleSpinBox::setSpecialValueText( txt );
+}
+
 QString QgsDoubleSpinBox::stripped( const QString &originalText ) const
 {
   //adapted from QAbstractSpinBoxPrivate::stripped
   //trims whitespace, prefix and suffix from spin box text
   QString text = originalText;
   if ( specialValueText().isEmpty() || text != specialValueText() )
   {
+    // Strip SPECIAL_TEXT_WHEN_EMPTY
+    if ( text.contains( SPECIAL_TEXT_WHEN_EMPTY ) )
```

Comment:
`==`?

Reply:
No: the problem that is addressed here is that when the widget value is null and the user enters a text, the text gets appended/prepended to the invisible character and the value is not correctly converted to double/int, this is also checked in the tests, so we really want to replace any invisible char in any position in the test.

Reply:
:+1: 

Sidenote: at some point I'll create a number-optimized QLineEdit with proper placeholder text and int/double properties. Noone needs the arrows on spinboxes anyway. And the only thing they reliably get right is data corruption because they grab the mouse wheel.

:skull_and_crossbones:  death to spinboxes :skull_and_crossbones:

Reply:
Oh, and they could even make copy/paste with thousand separators work as expected.